# Домашнее задание к лекции "Базовые понятия статистики"

### Задание 1. Базовое изучение

In [1]:
import pandas as pd
import numpy as np
import re

data = pd.read_csv('horse_data.csv', header=None,  na_values='?')
data.head()

0   1        2     3      4     5    6    7    8    9   ...    18    19  \
0  2.0   1   530101  38.5   66.0  28.0  3.0  3.0  NaN  2.0  ...  45.0   8.4   
1  1.0   1   534817  39.2   88.0  20.0  NaN  NaN  4.0  1.0  ...  50.0  85.0   
2  2.0   1   530334  38.3   40.0  24.0  1.0  1.0  3.0  1.0  ...  33.0   6.7   
3  1.0   9  5290409  39.1  164.0  84.0  4.0  1.0  6.0  2.0  ...  48.0   7.2   
4  2.0   1   530255  37.3  104.0  35.0  NaN  NaN  6.0  2.0  ...  74.0   7.4   

    20   21   22  23     24  25  26  27  
0  NaN  NaN  2.0   2  11300   0   0   2  
1  2.0  2.0  3.0   2   2208   0   0   2  
2  NaN  NaN  1.0   2      0   0   0   1  
3  3.0  5.3  2.0   1   2208   0   0   1  
4  NaN  NaN  2.0   2   4300   0   0   2  

[5 rows x 28 columns]

In [2]:
column_names = dict()
with open('horse_data.names') as f:
    for line in f:
        find_list = re.findall(r'(\d{1,2}):\s+((\w+.\s+)+)', line)
        if find_list:
            column_names[int(find_list[0][0])-1] = find_list[0][1].strip()
            
column_names

{0: 'surgery?',
 1: 'Age',
 2: 'Hospital Number',
 3: 'rectal temperature',
 4: 'pulse',
 5: 'respiratory rate',
 6: 'temperature of extremities',
 7: 'peripheral pulse',
 8: 'mucous membranes',
 9: 'capillary refill time',
 10: 'pain',
 11: 'peristalsis',
 12: 'abdominal distension',
 13: 'nasogastric tube',
 14: 'nasogastric reflux',
 15: 'nasogastric reflux PH',
 16: 'rectal examination',
 17: 'abdomen',
 18: 'packed cell volume',
 19: 'total protein',
 20: 'abdominocentesis appearance',
 21: 'abdomcentesis total protein',
 22: 'outcome',
 23: 'surgical lesion?',
 26: 'type of lesion',
 27: 'cp_data'}

In [3]:
# 8 столбцов для дальнейшего изучения
columns_chosen_no = [0, 1, 2, 3, 4, 5, 15, 22]
columns_chosen = [column_names[i] for i in columns_chosen_no]
columns_chosen

['surgery?',
 'Age',
 'Hospital Number',
 'rectal temperature',
 'pulse',
 'respiratory rate',
 'nasogastric reflux PH',
 'outcome']

In [4]:
columns_to_drop = list(set(range(0,28)) - set(columns_chosen_no))
data.drop(data.columns[columns_to_drop], axis='columns', inplace = True)
data.columns = columns_chosen
data

surgery?  Age  Hospital Number  rectal temperature  pulse  \
0         2.0    1           530101                38.5   66.0   
1         1.0    1           534817                39.2   88.0   
2         2.0    1           530334                38.3   40.0   
3         1.0    9          5290409                39.1  164.0   
4         2.0    1           530255                37.3  104.0   
..        ...  ...              ...                 ...    ...   
295       1.0    1           533886                 NaN  120.0   
296       2.0    1           527702                37.2   72.0   
297       1.0    1           529386                37.5   72.0   
298       1.0    1           530612                36.5  100.0   
299       1.0    1           534618                37.2   40.0   

     respiratory rate  nasogastric reflux PH  outcome  
0                28.0                    NaN      2.0  
1                20.0                    NaN      3.0  
2                24.0                    NaN      1.0  
3                84.0                    5.0      2.0  
4                35.0                    NaN      2.0  
..                ...                    ...      ...  
295              70.0                    NaN      3.0  
296              24.0                    NaN      3.0  
297              30.0                    NaN      2.0  
298              24.0                    NaN      1.0  
299              20.0                    NaN      3.0  

[300 rows x 8 columns]

In [5]:
data.describe()

surgery?         Age  Hospital Number  rectal temperature  \
count  299.000000  300.000000     3.000000e+02          240.000000   
mean     1.397993    1.640000     1.085889e+06           38.167917   
std      0.490305    2.173972     1.529801e+06            0.732289   
min      1.000000    1.000000     5.184760e+05           35.400000   
25%      1.000000    1.000000     5.289040e+05           37.800000   
50%      1.000000    1.000000     5.303055e+05           38.200000   
75%      2.000000    1.000000     5.347275e+05           38.500000   
max      2.000000    9.000000     5.305629e+06           40.800000   

            pulse  respiratory rate  nasogastric reflux PH     outcome  
count  276.000000        242.000000              53.000000  299.000000  
mean    71.913043         30.417355               4.707547    1.551839  
std     28.630557         17.642231               1.982311    0.737187  
min     30.000000          8.000000               1.000000    1.000000  
25%     48.000000         18.500000               3.000000    1.000000  
50%     64.000000         24.500000               5.000000    1.000000  
75%     88.000000         36.000000               6.500000    2.000000  
max    184.000000         96.000000               7.500000    3.000000

In [6]:
for param in ['rectal temperature', 'pulse', 'respiratory rate', 'nasogastric reflux PH']:
    print('\t', '\033[1m', param.upper(), '\033[0m')
    print('Максимальное значение: ', data[param].max()) 
    print('Минимальное значение: ', data[param].min()) 
    print('Среднее значение: ',  round(data[param].mean(),2))
    range_ = data[param].max() - data[param].min()
    print('Размах: ', round(range_,2))
    print('Значение моды: ', data[param].mode()[0])
    print('Значение медианы: ', data[param].median())
    print('СКО: ', round(data[param].std(),2))

	  RECTAL TEMPERATURE 
Максимальное значение:  40.8
Минимальное значение:  35.4
Среднее значение:  38.17
Размах:  5.4
Значение моды:  38.0
Значение медианы:  38.2
СКО:  0.73
	  PULSE 
Максимальное значение:  184.0
Минимальное значение:  30.0
Среднее значение:  71.91
Размах:  154.0
Значение моды:  48.0
Значение медианы:  64.0
СКО:  28.63
	  RESPIRATORY RATE 
Максимальное значение:  96.0
Минимальное значение:  8.0
Среднее значение:  30.42
Размах:  88.0
Значение моды:  20.0
Значение медианы:  24.5
СКО:  17.64
	  NASOGASTRIC REFLUX PH 
Максимальное значение:  7.5
Минимальное значение:  1.0
Среднее значение:  4.71
Размах:  6.5
Значение моды:  2.0
Значение медианы:  5.0
СКО:  1.98


### Описание результатов:
1. Средняя температура 38.2 не равна нормальному значению (37,8), так как работаем с больными лошадьми
2. Наиболее часто встречающееся значение пульса 48 (ударов в минуту). Судя по размаху (154) и медиане (64), пульс выходит далеко за рамки нормальных значений (30-40). Также означает заболевания лошадей.
3. Нормальное значение частоты дыхания от 8 до 10. Среднее (30,4) и медианное (24) значения не совпадают, имеется большой размах (88). Использование данной метрики сомнительно из-за большого разброса.
4. Водородный показатель pH гастроэзофагеального рефлюкса. Нормальное значение от 3 до 4. Возможные значения от 0 до 14. Нейтральный рН равен 7. Данные удовлетворяют критериям.

### Задание 2. Работа с выбросами

In [7]:
#найдем выбросы через межкваритльный размах
bounds = dict()
for param in ['rectal temperature','pulse', 'respiratory rate', 'nasogastric reflux PH']:
    print('\t', '\033[1m', param.upper(), '\033[0m')
    q1 = data[param].quantile(0.25)
    q3 = data[param].quantile(0.75)
    iqr = q3 - q1
    lower_bound = q1 - (1.5 * iqr) 
    upper_bound = q3 + (1.5 * iqr)
    bounds[param] = (lower_bound, upper_bound)
    print('Нижняя граница', round(lower_bound,2))
    print('Верхняя граница', round(upper_bound,2))
    print('Среднее значение с выбросами',data[param].mean())
    print('Среднее значение без выбросов',data[(data[param] > lower_bound) & (data[param] < upper_bound)][param].mean())
#     remove_outliers = remove_outliers[remove_outliers[param].between(lower_bound, upper_bound, inclusive=True)]

	  RECTAL TEMPERATURE 
Нижняя граница 36.75
Верхняя граница 39.55
Среднее значение с выбросами 38.16791666666669
Среднее значение без выбросов 38.169026548672576
	  PULSE 
Нижняя граница -12.0
Верхняя граница 148.0
Среднее значение с выбросами 71.91304347826087
Среднее значение без выбросов 70.25830258302582
	  RESPIRATORY RATE 
Нижняя граница -7.75
Верхняя граница 62.25
Среднее значение с выбросами 30.417355371900825
Среднее значение без выбросов 26.715555555555557
	  NASOGASTRIC REFLUX PH 
Нижняя граница -2.25
Верхняя граница 11.75
Среднее значение с выбросами 4.707547169811321
Среднее значение без выбросов 4.707547169811321


In [8]:
bounds

{'rectal temperature': (36.74999999999999, 39.550000000000004),
 'pulse': (-12.0, 148.0),
 'respiratory rate': (-7.75, 62.25),
 'nasogastric reflux PH': (-2.25, 11.75)}

### Средняя температура 
с выбросами и без них практически не изменилась, поэтому данные выбросы оставляем.

## Пульс
* Верхняя граница пульса 148 уд/мин
* нормальные значения для взрослых лошадей 30 - 40 уд/мин
* нормальные значения для жеребцов 80 - 100 уд/мин
* усиленное сердцебиения может говорить о болезненных поражениях или о циркуляторном шоке, который испытывает лошадь
#### пробуем убрать выбросы

In [9]:
remove_outliers = data[data['pulse'].between(bounds['pulse'][0], bounds['pulse'][1], inclusive=True)].sort_values('pulse')
remove_outliers

surgery?  Age  Hospital Number  rectal temperature  pulse  \
130       1.0    1           528964                37.8   30.0   
232       1.0    1          5262543                38.5   30.0   
203       2.0    1           529685                37.2   36.0   
276       1.0    1           534787                38.0   36.0   
242       1.0    1           530354                 NaN   36.0   
..        ...  ...              ...                 ...    ...   
148       1.0    1           529272                38.3  132.0   
135       1.0    9          5287179                38.1  136.0   
44        1.0    1           535407                35.4  140.0   
103       1.0    9          5292489                38.0  140.0   
39        1.0    9          5277409                39.2  146.0   

     respiratory rate  nasogastric reflux PH  outcome  
130              12.0                    NaN      2.0  
232              18.0                    NaN      1.0  
203               9.0                    NaN      1.0  
276              16.0                    2.0      3.0  
242              32.0                    4.0      2.0  
..                ...                    ...      ...  
148               NaN                    6.2      1.0  
135              48.0                    4.4      2.0  
44               24.0                    NaN      3.0  
103              68.0                    NaN      1.0  
39               96.0                    NaN      2.0  

[271 rows x 8 columns]

In [10]:
# Какие выбросы убрали:
outliers = pd.concat([data, remove_outliers]).drop_duplicates(keep=False)
print('Количество пропусков:', outliers['pulse'].isnull().sum())
outliers[outliers['pulse'] > 0]

Количество пропусков: 20


surgery?  Age  Hospital Number  rectal temperature  pulse  \
3         1.0    9          5290409                39.1  164.0   
41        2.0    9          5288249                39.0  150.0   
55        1.0    9          5282839                38.6  160.0   
255       1.0    9          5294539                38.8  184.0   
275       1.0    9          5297159                38.8  150.0   

     respiratory rate  nasogastric reflux PH  outcome  
3                84.0                    5.0      2.0  
41               72.0                    NaN      1.0  
55               20.0                    NaN      2.0  
255              84.0                    NaN      2.0  
275              50.0                    NaN      2.0

## Частота дыхания (respiratory rate)
* Нормальное значение частоты дыхания от 8 до 10. Для жеребцов 20 - 40

In [11]:
bounds['respiratory rate']

(-7.75, 62.25)

In [12]:
remove_outliers = data[data['respiratory rate'].between(bounds['respiratory rate'][0], bounds['respiratory rate'][1], inclusive=True)].sort_values('respiratory rate')
remove_outliers

surgery?  Age  Hospital Number  rectal temperature  pulse  \
190       1.0    1           528742                37.1   40.0   
203       2.0    1           529685                37.2   36.0   
140       1.0    1           529960                38.1   44.0   
89        2.0    1           535381                38.3   42.0   
162       2.0    1           535176                39.5   60.0   
..        ...  ...              ...                 ...    ...   
220       1.0    1           535029                39.2   88.0   
20        1.0    1           530157                39.9   72.0   
23        1.0    9           534998                38.3  130.0   
185       1.0    1           534833                38.5  120.0   
248       1.0    1           527709                38.0  108.0   

     respiratory rate  nasogastric reflux PH  outcome  
190               8.0                    NaN      1.0  
203               9.0                    NaN      1.0  
140               9.0                    NaN      1.0  
89               10.0                    NaN      1.0  
162              10.0                    NaN      1.0  
..                ...                    ...      ...  
220              58.0                    NaN      3.0  
20               60.0                    NaN      1.0  
23               60.0                    NaN      1.0  
185              60.0                    NaN      1.0  
248              60.0                    NaN      1.0  

[225 rows x 8 columns]

In [13]:
# Какие выбросы убрали:
outliers = pd.concat([data, remove_outliers]).drop_duplicates(keep=False)
print('Количество пропусков:', outliers['respiratory rate'].isnull().sum())
outliers[outliers['respiratory rate'] > 0]

Количество пропусков: 52


surgery?  Age  Hospital Number  rectal temperature  pulse  \
3         1.0    9          5290409                39.1  164.0   
39        1.0    9          5277409                39.2  146.0   
41        2.0    9          5288249                39.0  150.0   
82        1.0    9          5290759                38.1  100.0   
84        1.0    1           529849                37.8   60.0   
103       1.0    9          5292489                38.0  140.0   
106       1.0    1           528668                38.3   52.0   
120       1.0    1           528682                39.4   54.0   
125       1.0    1           528461                38.0   42.0   
186       1.0    1          5278331                39.3   64.0   
208       1.0    1           535338                37.8   88.0   
229       1.0    9           534597                38.5  120.0   
244       1.0    9          5291409                38.2  124.0   
255       1.0    9          5294539                38.8  184.0   
265       2.0    1           528590                 NaN   56.0   
269       1.0    1           527883                39.3   64.0   
295       1.0    1           533886                 NaN  120.0   

     respiratory rate  nasogastric reflux PH  outcome  
3                84.0                    5.0      2.0  
39               96.0                    NaN      2.0  
41               72.0                    NaN      1.0  
82               80.0                    NaN      1.0  
84               80.0                    5.5      1.0  
103              68.0                    NaN      1.0  
106              96.0                    NaN      1.0  
120              66.0                    NaN      1.0  
125              68.0                    NaN      1.0  
186              90.0                    NaN      1.0  
208              80.0                    NaN      3.0  
229              70.0                    NaN      1.0  
244              88.0                    NaN      1.0  
255              84.0                    NaN      2.0  
265              68.0                    NaN      3.0  
269              90.0                    6.5      1.0  
295              70.0                    NaN      3.0

### из описания
измерение частоты дыхания часто дает большую погрешность, в добавок имеется большое количество пропусков, в дальнейшнем столбец rectal temperature возможно придется удалить.

### nasogastric reflux PH

In [14]:
bounds['nasogastric reflux PH']

(-2.25, 11.75)

* Водородный показатель pH гастроэзофагеального рефлюкса в данной выборке имеет максимальное значение 7,5. Верхняя граница через межкваритльный размах получилась 11,75. Поэтому в данной выборке выбросов нет.

### Задание 3. Работа с пропусками

Рассчитать количество выбросов для всех выбранных столбцов. Принять и обосновать решение о методе работы с пропусками по каждому столбцу, сформировать датафрейм, в котором пропуски будут отсутствовать.

In [15]:
for col in data.columns:
    pct_missing = data[col].isnull().mean()
    print(f'{col} - {pct_missing :.1%}')

surgery? - 0.3%
Age - 0.0%
Hospital Number - 0.0%
rectal temperature - 20.0%
pulse - 8.0%
respiratory rate - 19.3%
nasogastric reflux PH - 82.3%
outcome - 0.3%


1. Столбец surgery (была ли хирургическая операция) является категориальным, может иметь только 2 значения, 1 - была операция, 2 - лечили без операций. Имеет 0,3 % пропусков. Данные пропуски удаляем, так как нет возможности узнать эту информацию.
2. Столбец Age (возраст) является категориальным, может иметь только 2 значения. Не имеет пропусков.
3. Столбец Hospital Number (номер присвоенный заболевшей лошади) является категориальным. Не имеет пропусков.
4. Столбец rectal temperature (ректальная температура) содержит 20 % пропусков.
5. Столбец pulse - 8.0 %. Пропуски пульс заполняем средним значением, раз его не измеряли, значит не так важен был.
6. respiratory rate - 19.3 %. Данная характеристика имеет большую погрешность, и ее полезность подвергается сомнению, поэтому логично этот столбец удалить.
7. nasogastric reflux PH - 82.3 %. очень много пропусков. Неизвестно как их корректно заполнить. Поэтому надо удалить полностью данный столбец. 
8. outcome - 0.3 %. Строки с незаполненными данными о финальном исходе удаляем, так как пропусков всего 0,3%.

### Начнем со столбца nasogastric reflux PH, так как в нем больше всего пропусков. Удалим его

In [16]:
data_drop_missings = data.drop(['nasogastric reflux PH'], axis=1)

In [17]:
data_drop_missings

surgery?  Age  Hospital Number  rectal temperature  pulse  \
0         2.0    1           530101                38.5   66.0   
1         1.0    1           534817                39.2   88.0   
2         2.0    1           530334                38.3   40.0   
3         1.0    9          5290409                39.1  164.0   
4         2.0    1           530255                37.3  104.0   
..        ...  ...              ...                 ...    ...   
295       1.0    1           533886                 NaN  120.0   
296       2.0    1           527702                37.2   72.0   
297       1.0    1           529386                37.5   72.0   
298       1.0    1           530612                36.5  100.0   
299       1.0    1           534618                37.2   40.0   

     respiratory rate  outcome  
0                28.0      2.0  
1                20.0      3.0  
2                24.0      1.0  
3                84.0      2.0  
4                35.0      2.0  
..                ...      ...  
295              70.0      3.0  
296              24.0      3.0  
297              30.0      2.0  
298              24.0      1.0  
299              20.0      3.0  

[300 rows x 7 columns]

### Далее: столбец respiratory rate. Ввиду больших погрешностей, столбец удаляем

In [18]:
data_drop_missings.drop(['respiratory rate'], axis=1, inplace=True)

### Далее: столбец surgery, а и заодно и outcome

In [19]:
# строки, в которых нет значения столбца surgery?
data_drop_missings.loc[pd.isnull(data_drop_missings['surgery?']), :]

surgery?  Age  Hospital Number  rectal temperature  pulse  outcome
132       NaN    1           534572                38.0   48.0      NaN

In [20]:
data_drop_missings.drop(data_drop_missings.loc[pd.isnull(data_drop_missings['surgery?']), :].index, inplace = True)

In [21]:
data_drop_missings 

surgery?  Age  Hospital Number  rectal temperature  pulse  outcome
0         2.0    1           530101                38.5   66.0      2.0
1         1.0    1           534817                39.2   88.0      3.0
2         2.0    1           530334                38.3   40.0      1.0
3         1.0    9          5290409                39.1  164.0      2.0
4         2.0    1           530255                37.3  104.0      2.0
..        ...  ...              ...                 ...    ...      ...
295       1.0    1           533886                 NaN  120.0      3.0
296       2.0    1           527702                37.2   72.0      3.0
297       1.0    1           529386                37.5   72.0      2.0
298       1.0    1           530612                36.5  100.0      1.0
299       1.0    1           534618                37.2   40.0      3.0

[299 rows x 6 columns]

In [22]:
for col in data_drop_missings.columns:
    pct_missing = data_drop_missings[col].isnull().mean()
    print(f'{col} - {pct_missing :.1%}')

surgery? - 0.0%
Age - 0.0%
Hospital Number - 0.0%
rectal temperature - 20.1%
pulse - 8.0%
outcome - 0.0%


### Столбец rectal temperature

In [23]:
# строки с пропусками в столбце rectal temperature
data_drop_missings.loc[pd.isnull(data_drop_missings['rectal temperature']), :].head(10)

surgery?  Age  Hospital Number  rectal temperature  pulse  outcome
5        2.0    1           528355                 NaN    NaN      1.0
7        1.0    1           529607                 NaN   60.0      2.0
8        2.0    1           530051                 NaN   80.0      3.0
16       1.0    9          5301219                 NaN  128.0      2.0
28       1.0    1          5279442                 NaN    NaN      2.0
34       1.0    1           529796                 NaN  100.0      1.0
35       2.0    1           528812                 NaN  104.0      3.0
40       1.0    1           529498                 NaN   88.0      2.0
43       1.0    1           534069                 NaN  120.0      3.0
45       2.0    1           529827                 NaN  120.0      2.0

#### Заполняем пропуски в температуре на медианное значение в зависимости от исхода (outcome)

In [24]:
data_drop_missings['rectal temperature'].fillna(data.groupby('outcome')['rectal temperature'].transform('median'), inplace=True)

In [25]:
data_drop_missings

surgery?  Age  Hospital Number  rectal temperature  pulse  outcome
0         2.0    1           530101               38.50   66.0      2.0
1         1.0    1           534817               39.20   88.0      3.0
2         2.0    1           530334               38.30   40.0      1.0
3         1.0    9          5290409               39.10  164.0      2.0
4         2.0    1           530255               37.30  104.0      2.0
..        ...  ...              ...                 ...    ...      ...
295       1.0    1           533886               38.05  120.0      3.0
296       2.0    1           527702               37.20   72.0      3.0
297       1.0    1           529386               37.50   72.0      2.0
298       1.0    1           530612               36.50  100.0      1.0
299       1.0    1           534618               37.20   40.0      3.0

[299 rows x 6 columns]

In [26]:
for col in data_drop_missings.columns:
    pct_missing = data_drop_missings[col].isnull().mean()
    print(f'{col} - {pct_missing :.1%}')

surgery? - 0.0%
Age - 0.0%
Hospital Number - 0.0%
rectal temperature - 0.0%
pulse - 8.0%
outcome - 0.0%


### Столбец pulse

In [27]:
# строки с пропусками в столбце pulse
data_drop_missings.loc[pd.isnull(data_drop_missings['pulse']), :].head()

surgery?  Age  Hospital Number  rectal temperature  pulse  outcome
5        2.0    1           528355                38.2    NaN      1.0
28       1.0    1          5279442                38.1    NaN      2.0
52       2.0    1           529483                38.2    NaN      1.0
56       1.0    1           528872                38.2    NaN      1.0
58       1.0    1           528298                38.1    NaN      2.0

#### Заполняем пропуски в пульсе на моду (48 ударов в минуту, немного выше нормальных значений, но это приемлимо, так как лошади нездоровы)

In [28]:
data_drop_missings['pulse'].fillna(data['pulse'].mode()[0], inplace=True)

In [29]:
for col in data_drop_missings.columns:
    pct_missing = data_drop_missings[col].isnull().mean()
    print(f'{col} - {pct_missing :.1%}')

surgery? - 0.0%
Age - 0.0%
Hospital Number - 0.0%
rectal temperature - 0.0%
pulse - 0.0%
outcome - 0.0%


In [30]:
# Итоговый датафрейм без пропусков
data_drop_missings

surgery?  Age  Hospital Number  rectal temperature  pulse  outcome
0         2.0    1           530101               38.50   66.0      2.0
1         1.0    1           534817               39.20   88.0      3.0
2         2.0    1           530334               38.30   40.0      1.0
3         1.0    9          5290409               39.10  164.0      2.0
4         2.0    1           530255               37.30  104.0      2.0
..        ...  ...              ...                 ...    ...      ...
295       1.0    1           533886               38.05  120.0      3.0
296       2.0    1           527702               37.20   72.0      3.0
297       1.0    1           529386               37.50   72.0      2.0
298       1.0    1           530612               36.50  100.0      1.0
299       1.0    1           534618               37.20   40.0      3.0

[299 rows x 6 columns]

In [31]:
data_drop_missings = data.drop(['nasogastric reflux PH'], axis=1)

In [32]:
# Заполняем пропуски в пульсе на медианное значение в зависимости от наличия/отсутствия операции и температуры. P.S. не сработало почему-то
data_drop_missings['pulse'].fillna(data.groupby(['rectal temperature', 'surgery?'])['pulse'].transform('median'), inplace=True)

In [33]:
data_drop_missings.loc[pd.isnull(data_drop_missings['pulse']), :].head()

surgery?  Age  Hospital Number  rectal temperature  pulse  \
5        2.0    1           528355                 NaN    NaN   
28       1.0    1          5279442                 NaN    NaN   
52       2.0    1           529483                 NaN    NaN   
56       1.0    1           528872                 NaN    NaN   
58       1.0    1           528298                 NaN    NaN   

    respiratory rate  outcome  
5                NaN      1.0  
28               NaN      2.0  
52               NaN      1.0  
56               NaN      1.0  
58              20.0      2.0

In [34]:
data.groupby(['rectal temperature', 'surgery?'])['pulse'].median()

rectal temperature  surgery?
35.4                1.0         140.0
36.0                2.0          42.0
36.1                2.0          88.0
36.4                1.0          98.0
36.5                1.0          89.0
                                ...  
39.7                1.0         100.0
39.9                1.0          72.0
40.0                2.0          78.0
40.3                2.0         114.0
40.8                1.0          72.0
Name: pulse, Length: 61, dtype: float64